README.adoc
= Data4Life Auth SDK
:link-repository: https://github.com/d4l-data4life/hc-auth-sdk-kmp
:project-version: 1.17.0
:doctype: article
:!showtitle:
:toc: macro
:toclevels: 2
:toc-title:
:icons: font
:imagesdir: assets/images
ifdef::env-github[]
:warning-caption: :warning:
:caution-caption: :fire:
:important-caption: :exclamation:
:note-caption: :paperclip:
:tip-caption: :bulb:
endif::[]

++++
<div align="center">
    <p><!-- PROJECT LOGO -->
        <a href="https://github.com/d4l-data4life/hc-auth-sdk-kmp">
            <img src="assets/images/d4l-logo.svg" width="150"/>
        </a>
    </p>
    <p><!-- PROJECT TITLE -->
        <h3>Data4Life Auth SDK</h3>
    </p>
    <p><!-- PROJECT DESCRIPTION -->
        Authorization SDK for communicating with PHDC using OAuth 2.0
    </p>
    <p><!-- PROJECT DOCUMENTATION -->
        <a href="README.adoc"><strong>Explore the docs »</strong></a>
    </p>
    <p><!-- PROJECT ISSUES/FEATURES -->
        <a href="https://github.com/d4l-data4life/hc-auth-sdk-kmp/issues">Report Bug</a>
        ·
        <a href="https://github.com/d4l-data4life/hc-auth-sdk-kmp/issues">Request Feature</a>
    </p>
    <p><!-- PROJECT BADGES see badges.adoc how to change them -->
        <a href="https://github.com/d4l-data4life/hc-auth-sdk-kmp/releases">
            <img src="assets/images/badge-release-latest.svg" alt="Latest release"/>
        </a>
        <a href="https://github.com/d4l-data4life/hc-auth-sdk-kmp/actions">
            <img src="https://github.com/d4l-data4life/hc-auth-sdk-kmp/actions/workflows/d4l-ci-latest-version.yml/badge.svg" alt="D4L CI - Latest Version"/>
        </a>
        <a>
            <img src="assets/images/badge-platform-support.svg" alt="Platform"/>
        </a>
        <a href="LICENSE">
            <img src="assets/images/badge-license.svg" alt="License"/>
        </a>
    </p>
</div>
++++

[discrete]
=== Table Of Contents

toc::[]

== About The Project

Authorization SDK for communicating with PHDC using link:https://tools.ietf.org/html/rfc6749[OAuth 2.0]. For Android it implements link:https://tools.ietf.org/html/rfc8252[RFC 8252 - OAuth 2.0 for Native Apps] using link:https://github.com/openid/AppAuth-Android[AppAuth]. Java is supporting the OAuth 2.0 Code Grant using link:https://github.com/scribejava/scribejava[ScribeJava].

=== Features

* Android: OAuth 2.0 - Native App flow
* Java: OAuth 2.0 - Code Grant flow

=== Built With

* link:https://kotlinlang.org/docs/reference/mpp-intro.html[Kotlin Multiplatform]

== Getting Started

Instructions how to get software up and running.

=== Prerequisites

* link:https://developer.android.com/studio#downloads[Android Studio - Chipmunk - 2021.2.1]
* link:https://adoptopenjdk.net/[Java 11] | link:https://developer.android.com/studio/write/java8-support[Limitations of Java language features and APIs] | https://jakewharton.com/d8-library-desugaring/[Desugaring]
* link:https://kotlinlang.org/[Kotlin 1.7.10]
* Android 6.0 (API 23) to Android 13 (API 33)

=== Installation

A step by step series of examples that tell you how to get it running.

==== Consume GitHub Packages

We use GitHub Packages to distribute the SDK. In order to consume our dependencies you need to generate a GitHub Personal Access Token. Please follow the how to link:https://docs.github.com/en/packages/learn-github-packages/introduction-to-github-packages#authenticating-to-github-packages[authenticate to GitHub Packages, window="_blank"].

NOTICE:: You need to have _read_ access to this repository and generate a https://github.com/settings/tokens/new/[personal access token] with `repo` and `read:packages` scope.

===== Token

The token needs to be made available.

. Add `gpr.user = {GitHub username}` and `gpr.key = {GitHub Personal Access Token}` to your global Gradle properties `~/.gradle/gradle.properties`

    gpr.user=github-username
    gpr.key=github-token

. Or add following environment variables `PACKAGE_REGISTRY_USERNAME={GitHub username}` and `PACKAGE_REGISTRY_TOKEN={GitHub Personal Access Token}`

===== Setup Maven Repository

Add the following maven repository configuration to your root build.gradle:

[source, Gradle]
----
    allprojects {
        repositories {
            ...
            maven {
                url = URI("https://maven.pkg.github.com/d4l-data4life/hc-auth-sdk-kmp")
                credentials {
                    username = project.findProperty("gpr.user") as String? ?: System.getenv("PACKAGE_REGISTRY_USERNAME")
                    password = project.findProperty("gpr.key") as String? ?: System.getenv("PACKAGE_REGISTRY_TOKEN")
                }
            }
        }
    }
----

==== Java

Add following dependencies to your app build.gradle.

----
dependencies {
    implementation "care.data4life.hc-auth-sdk-kmp:auth-jvm:$version"
}
----

==== Android

----
dependencies {
    implementation "care.data4life.hc-auth-sdk-kmp:auth-android:$version"
}
----

In case you target Android 11+ and experience a `ActivityNotFoundException`. This is caused by the changed link:https://developer.android.com/about/versions/11/privacy/package-visibility[package visibility in Android 11]. Please ensure that the following is part of your merged `AndroidManifest.xml`. In case not, add the following to your `AndroidManifest.xml` file:

----
<manifest>
    <application>
        ...
    </application>

    <queries>
        <intent>
            <action android:name="android.intent.action.VIEW" />
            <category android:name="android.intent.category.BROWSABLE" />
            <data android:scheme="https" />
        </intent>
    </queries>

</manifest>
----

== Usage

The authorization request is started with the following code sample which launches an link:https://developer.chrome.com/docs/android/custom-tabs/[Custom Tab]
to present the login screen to the user for authentication.

[source]
----
Intent authIntent = Data4LifeClient.getInstance().getLoginIntent(null);
startActivityForResult(authIntent,requestCode);
----

Once the user finished, canceled or an error occurred during the browser login, the SDK sends the response to the
integrator application in the `onActivityResult(…)` method.

The authorization results can be the following:

- `Activity.RESULT_OK` – when the login is successful.
- `Activity.RESULT_CANCELED` – when the login fails, with additional payload in `data`, as shown in the example.

To end the login you need to call `finishLogin(authData, callback)` with the intent received in `onActivityResult(…)`.
The callback will indicate if the authorization successfully finished.

[source]
----
@Override protected void onActivityResult(int requestCode, int resultCode, Intent data) {
    super.onActivityResult(requestCode, resultCode, data);
    if (requestCode == GC_AUTH) {
        if (resultCode == RESULT_OK) {
            client.finishLogin(data,callback);
        } else if (data.getExtras() != null) {
            if (data.getExtras().containsKey("error")) {
                Snackbar.make(mRootCL, "Failed to log in to Data4Life", Snackbar.LENGTH_SHORT).show();
            } else if (data.getExtras().containsKey("canceled")) {
                Snackbar.make(mRootCL, "User canceled authorization request", Snackbar.LENGTH_SHORT).show();
            }
        }
    }
}
----

== Roadmap

This project is work in progress. We are working on adding more functionality, guidelines, documentation and other improvements.

Also see the open link:{link-repository}/issues[issues] for a list of proposed features and known issues.

== Changelog

See link:CHANGELOG.adoc[changelog]

== Versioning

We use http://semver.org/[Semantic Versioning] as a guideline for our versioning.

Releases use this format: `{major}.{minor}.{patch}`

* Breaking changes bump `{major}` and reset `{minor}` & `{patch}`
* Backward compatible changes bump `{minor}` and reset `{patch}`
* Bug fixes bump `{patch}`

== Contributing

You want to help or share a proposal? You have a specific problem? Read the following:

* link:CODE-OF-CONDUCT.adoc[Code of conduct] for details on our code of conduct.
* link:CONTRIBUTING.adoc[Contributing] for details about how to report bugs and propose features.
* link:DEVELOPING.adoc[Developing] for details about our development process and how to build and test the project.

== Copyright and License

Copyright (c) 2022 D4L data4life gGmbH / All rights reserved.

Please refer to our link:LICENSE[License] for further details.


CHANGELOG.adoc
= Changelog
:link-repository: https://github.com/d4l-data4life/hc-auth-sdk-kmp
:doctype: article
:toc: macro
:toclevels: 1
:toc-title:
:icons: font
:imagesdir: assets/images
ifdef::env-github[]
:warning-caption: :warning:
:caution-caption: :fire:
:important-caption: :exclamation:
:note-caption: :paperclip:
:tip-caption: :bulb:
endif::[]

All notable changes to this project will be documented in this file.

The format is based on http://keepachangelog.com/en/1.0.0/[Keep a Changelog]
and this project adheres to http://semver.org/spec/v2.0.0.html[Semantic Versioning].

Template could be found here: link:https://github.com/d4l-data4life/hc-readme-template/blob/main/TEMPLATE_CHANGELOG.adoc[CHANGELOG]

[discrete]
==== Versions

toc::[]
== link:{link-repository}/releases/latest[Unreleased]

See link:{link-repository}/compare/v1.17.0\...main[changeset]

=== Added

=== Changed

=== Deprecated

=== Removed

=== Fixed

=== Security

=== Bumped

=== Migration

== link:{link-repository}/releases/v1.17.0[1.17.0]

See link:{link-repository}/compare/v1.16.0\...v1.17.0[changeset]

=== Bumped

* Android Gradle Plugin 7.2.0 -> 7.2.2
* *_BREAKING_* Kotlin 1.6.21 -> 1.7.10
* Kotlin Coroutines 1.6.2 -> 1.6.3
* Kotlin Serialization 1.3.2 -> 1.3.3
* Kotlin DateTime 0.3.2 -> 0.4.0
* Android target SDK 32 -> 33
* HC Util 1.13.0 -> 1.14.0
* HC SecureStore SDK 1.15.0 -> 1.16.0

== link:{link-repository}/tag/v1.16.0[1.16.0]

See link:{link-repository}/compare/v1.15.0\..v.1.16.0[changeset]

=== Changed

* *_BREAKING_* login flow without LoginActivity. This requires to finishLogin from client side by calling `finishLogin`
* Android target SDK 31 -> 32

=== Removed

* LoginActivity
* Material dependency
* AndroidX ConstraintLayout dependency
* AndroidX AppCompat dependency
* AndroidX Browser dependency
* AndroidX Ktx dependency

=== Bumped

* Android Gradle Plugin 7.1.3 -> 7.2.0
* Kotlin 1.6.10 -> 1.6.21
* Kotlin Coroutines 1.6.0 -> 1.6.2
* Spotless Gradle plugin 6.4.2 -> 6.5.1
* Ktlint 0.44.0 -> 0.45.2

== link:{link-repository}/tag/v1.15.0[1.15.0]

See link:{link-repository}/compare/v1.14.0\...v1.15.0[changeset]

=== Removed

* HC Result

=== Bumped

* Gradle 6.9 -> 7.4.2
* *_BREAKING_* Kotlin 1.4.32 -> 1.6.10
* *_BREAKING_* Android Gradle Plugin 4.2.2 -> 7.1.3
* *_BREAKING_* Kotlin Coroutines 1.5.1 -> 1.6.0
* *_BREAKING_* Kotlin Serialization 1.2.2 -> 1.3.2
* *_BREAKING_* Kotlin DateTime 0.2.1 -> 0.3.2
* Tink 1.2.2 -> 1.4.0
* ScribeJava 6.1.0 -> 6.3.0
* HC Util SDK 1.8.0 -> 1.13.0
* HC SecureStore SDK 1.13.2 -> 1.15.0
* Android Ktx 1.3.2 -> 1.7.0
* Android AppCompat 1.3.0 -> 1.4.1
* Android Browser 1.2.0 -> 1.4.0
* Android ConstraintLayout 2.0.1 -> 2.1.3
* Android Material 1.4.0 -> 1.5.0
* Android AppAuth 0.10.0 -> 0.11.1
* Android Target SDK 30 -> 31

== link:{link-repository}/tag/v1.14.0[1.14.0]

See link:{link-repository}/compare/v1.13.2\...v1.14.0[changeset]

=== Added

* HC Result SDK 0.2.0

=== Changed

* project configuration to latest iteration

=== Fixed

* ActivityNotFoundException on Android 11+ when trying to login

=== Bumped

* Bump Gradle 6.9 -> 6.9.1
* Android Gradle Plugin 4.2.1 -> 4.2.2
* HC Util SDK 1.8.0 -> 1.10.0
* AndroidX Ktx 1.3.2 -> 1.6.0
* AndroidX AppCompat 1.3.0 -> 1.3.1
* AndroidX Browser 1.2.0 -> 1.3.0
* AppAuth from custom patched version -> official 0.10.0

== link:{link-repository}/releases/tag/v1.13.2[1.13.2]

Initial release taken out of link:https://github.com/d4l-data4life/hc-sdk-kmp/[HC SDK KMP].


buildSrc/src/main/kotlin/care/data4life/sdk/auth/LibraryConfig.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk.auth

object LibraryConfig {

    val publishConfig = PublishConfig
    val android = AndroidLibraryConfig

    const val group = "care.data4life.sdk.auth"
    const val name = "hc-auth-sdk-kmp"

    const val githubOwner = "d4l-data4life"
    const val githubRepository = "hc-auth-sdk-kmp"

    object PublishConfig {
        const val groupId = "care.data4life.hc-auth-sdk-kmp"

        const val description = "Provides Gradle scripts to harmonise Gradle project settings"
        const val year = "2021"

        // URL
        const val host = "github.com"
        const val path = "$githubOwner/$githubRepository"

        const val url = "https://$host/$path"

        // DEVELOPER
        const val developerId = "d4l-data4life"
        const val developerName = "D4L data4life gGmbH"
        const val developerEmail = "mobile@data4life.care"

        // LICENSE
        const val licenseName = "Private"
        const val licenseUrl = "$url/blob/main/LICENSE"
        const val licenseDistribution = "repo"

        // SCM
        const val scmUrl = "git://$host/$path.git"
        const val scmConnection = "scm:$scmUrl"
        const val scmDeveloperConnection = scmConnection
    }

    object AndroidLibraryConfig {
        const val minSdkVersion = 23
        const val compileSdkVersion = 30
        const val targetSdkVersion = 30

        const val resourcePrefix = "hc_auth_"
    }
}
